Initium-Odp/src/com/universeprojects/miniup/server/commands/CommandSwitchCharacter.java

```diff
@@ -38,15 +42,129 @@ public void run(Map<String, String> parameters) throws UserErrorMessage {
 			throw new UserErrorMessage(
 					"You cannot switch users if you're not logged into a user account. Are you just using a throwaway? If so, try converting the throwaway to a full account first.");
 		}
+		
+		CachedEntity currentCharacter = db.getCurrentCharacter();
 
 		// Get characterId to switch to
 		Long targetCharacterId;
+		
+		//This allows for hotkeys to loop through characters, either in alphabetical order or characters in the same party.
 		try {
-			targetCharacterId = Long.parseLong(parameters.get("characterId"));
-		} catch (NumberFormatException e) {
-			throw new UserErrorMessage("This character doesn't exist.");
-		}
+			int direction;
+			direction = Integer.parseInt(parameters.get("direction"));
+			
+			//This method will return all characters that we can switch to. Filters out zambies and dead characters right off the bat.
+			List<CachedEntity> characters = db.getAlphabetSortedValidCharactersByUser(user.getKey());
+			
+			switch(direction) {
+			
+			//UP in characterswitcher.
+			case 1:
+				for(int i = 0; i < characters.size(); i++) {
+					CachedEntity c = characters.get(i);
+					//here, we are searching for the currentcharacter, so we know its index within the sorted list.
+					if(!GameUtils.equals(c.getKey(), currentCharacter.getKey)) continue;					
+					
+					//if we're at the first element in the list
+					if(i == 0) {
+						targetCharacterId = characters.get(characters.size()).getId();
+						break;
+					}
+					
+					targetCharacterId = characters.get(i-1).getId();
+					break;
+				}
+				break;
+				
+			//DOWN in characterswitcher
+			case 2:
+				for(int i = 0; i < characters.size(); i++) {
+					CachedEntity c = characters.get(i);
+					//here, we are searching for the currentcharacter, so we know its index within the sorted list.
+					if(!GameUtils.equals(c.getKey(), currentCharacter.getKey)) continue;					
+					
+					//if we're at the last element in the list.
+					if(i == characters.size()) {
+						targetCharacterId = characters.get(0).getId();
+						break;
+					}
+					targetCharacterId = characters.get(i+1).getId();
+					break;
+				}
+				break;
+				
+			//Iterate through characters in the same party and userkey
+			case 3:
+				//here, we need to grab all the partied characters. then, filter by userkey.
+				List<CachedEntity> party = db.getParty(ds, currentCharacter);
+				if(party == null) throw new UserErrorMessage("You aren't in a party.");
```

Comment:
Same thing. Case fall through. Retrieve party, then check `direction` to determine logic for retrieving character.